clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -1574,7 +1583,11 @@ private boolean updateFetchPositions(final Timer timer) {
         try {
             CheckAndUpdatePositionsEvent checkAndUpdatePositionsEvent = new CheckAndUpdatePositionsEvent(calculateDeadlineMs(timer));
             wakeupTrigger.setActiveTask(checkAndUpdatePositionsEvent.future());
-            cachedSubscriptionHasAllFetchPositions = applicationEventHandler.addAndGet(checkAndUpdatePositionsEvent);
+            applicationEventHandler.add(checkAndUpdatePositionsEvent);
+            cachedSubscriptionHasAllFetchPositions = processBackgroundEvents(
```

Comment:
I'm afraid that processing background here events could bring undesired effects, mainly because it may include callbacks, which are totally unrelated to a call to consumer.position() for instance.

If there's a call to consumer.position, and a reconciliation starts in the background around the same time (ie. new partition assigned from the broker), the background will enqueue a CallbackNeededEvent, so we could end up actually running the callback here, as part of the call to position, which is not right.

I believe that we should not use `processBackgroundEvents` as a means of knowing that a specific event that we sent to the background failed. We should ensure that the event is completed exceptionally in the background instead. Wouldn't that work, and avoid mixing errors and callbacks?

Reply:
> I'm afraid that processing background here events could bring undesired effects, mainly because it may include callbacks, which are totally unrelated to a call to consumer.position() for instance.
If there's a call to consumer.position, and a reconciliation starts in the background around the same time (ie. new partition assigned from the broker), the background will enqueue a CallbackNeededEvent, so we could end up actually running the callback here, as part of the call to position, which is not right.

This is a preliminary thought: to prevent CallbackNeededEvent from being executed out of order, we could use a PriorityQueue to manage events based on their timestamps. This way, if an incorrect event is retrieved, it can be re-queued without disrupting the processing sequence.

> I believe that we should not use processBackgroundEvents as a means of knowing that a specific event that we sent to the background failed. We should ensure that the event is completed exceptionally in the background instead. Wouldn't that work, and avoid mixing errors and callbacks? (related info in the jira too)

I've been exploring the use of CompletableFuture to manage metadata exceptions in the new Kafka consumer. However, I've encountered some test failures, which may indicate differences in behavior compared to the classic consumer. This makes me question if this approach is the best way to refactor error handling in the new consumer.

Reply:
Hello @lianetm 
Let me first explain my understanding of the current issue and the approach I’m using.

When `consumer#position()` is called and reaches the line `applicationEventHandler.addAndGet(checkAndUpdatePositionsEvent);`, it gets blocked because `addAndGet` requires a response to proceed. However, in the current setup, the background thread converts errors into Error Events. This means the application thread won’t receive any errors if it hasn’t executed `processBackgroundEvents`.

>  we have a “direct” way of communicating the 2 threads

To address this, I’ve implemented a `CompletableFuture<Exception> metadataException` within the `NetworkClientDelegate `class specifically for storing errors that occur in the background thread. So in `NetworkClientDelegate#maybePropagateMetadataError`, if no error is present, I refresh the future. If there is an error, I call `metadataException.completeExceptionally(e);`.

The reason for this design is that when `ConsumerNetworkThread#processApplicationEvents` executes `checkAndUpdatePositionsEvent`, `TopicAuthorizationException` doesn’t immediately appear. Instead, it may take several iterations of `runOnce` before this error shows up. Without the future approach, this error can’t be propagated from the background thread to the OffsetsRequestManager. 

> Catch metadata errors on ConsumerNetworkThread.runOnce and completeExceptionally all the requests it has (requests returned by the managers poll here

I also catch the metadata error on this place, I think we can catch it when it also add to Error Event.
https://github.com/m1a2st/kafka/blob/51788cfa92881f29d95bf0ee197c6e5c8cd21ea6/clients/src/main/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegate.java#L156

> Notify managers of the metadata error too (ie. onMetadataERror), in case they have requests generated that did not make it to point 1, but still need to fail (ie. Commit request manager may have unsentOffsetCommitRequests/unsentOffsetFetchRequests. I don't think any other keeps requests without sending them to the network client but let's double check)

I also passing this error to the `OffsetsRequestManager#updateFetchPositions` when the process really get the event, It will passing the exception to the `OffsetsRequestManager` and let application thread know it is fail on `addAndGet` method.
https://github.com/m1a2st/kafka/blob/51788cfa92881f29d95bf0ee197c6e5c8cd21ea6/clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManager.java#L261

However, some tests still fail, likely due to behavioral differences from the classic consumer. I'm not sure that I miss something about this problem. Please let me which Im missing,  Thanks for your review and suggestion.